buildstockbatch/gcp/gcp.py

```diff
@@ -216,6 +241,160 @@ def push_image(self):
+
+        # Step 3: Define and run the GCP Batch job.
+        client = batch_v1.BatchServiceClient()
+
+        runnable = batch_v1.Runnable()
+        runnable.container = batch_v1.Runnable.Container()
+        # TODO: Use the docker image pushed earlier instead of this hardcoded image.
+        runnable.container.image_uri = 'us-central1-docker.pkg.dev/buildstockbatch-dev/buildstockbatch-docker/buildstockbatch'
+        runnable.container.entrypoint = '/bin/sh'
+
+        # Pass environment variables to each task
+        environment = batch_v1.Environment()
+        # TODO: What other env vars need to exist for run_job() below?
+        # BATCH_TASK_INDEX and BATCH_TASK_COUNT env vars are automatically made available.
+        environment.variables = {'JOB_ID': self.unique_job_id}
+        runnable.environment = environment
+
+        # TODO: Update to run batches of openstudio with something like "python3 -m buildstockbatch.gcp.gcp"
+        runnable.container.commands = [
+            "-c",
+            # Test script that checks whether openstudio is installed and writes to a file in GCS.
+            "mkdir /mnt/disks/share/${JOB_ID}; openstudio --help > /mnt/disks/share/${JOB_ID}/output_${BATCH_TASK_INDEX}.txt"
+        ]
+
+        # Mount GCS Bucket, so we can use it like a normal directory.
+        gcs_bucket = batch_v1.GCS(remote_path=self.bucket)
+        # Note: 'mnt/share/' is read-only, but 'mnt/disks/share' works
+        gcs_volume = batch_v1.Volume(
+            gcs = gcs_bucket,
+            mount_path = '/mnt/disks/share',
+        )
+
+        # TODO: Allow specifying resources from the project YAML file, plus pick better defaults.
+        resources = batch_v1.ComputeResource(
+            cpu_milli = 1000,
+            memory_mib = 1,
+        )
+
+        task = batch_v1.TaskSpec(
+            runnables = [runnable],
+            volumes = [gcs_volume],
+            compute_resource = resources,
+            # TODO: Confirm what happens if this fails repeatedly, or for only some tasks, and document it.
+            max_retry_count = 2,
+            # TODO: How long does this timeout need to be?
+            max_run_duration = '60s'
+        )
+
+        # How many of these tasks to run.
+        # TODO: determine this count above, when splitting up samples.
+        task_count = 3
+        group = batch_v1.TaskGroup(
+            task_count = task_count,
+            task_spec = task,
+        )
+
+        # Specify types of VMs to run on
+        # TODO: look into best default machine type for running OpenStudio, but also allow
+        # changing via the project config. https://cloud.google.com/compute/docs/machine-types
+        policy = batch_v1.AllocationPolicy.InstancePolicy(
+            machine_type = 'e2-standard-2',
+        )
+        instances = batch_v1.AllocationPolicy.InstancePolicyOrTemplate(policy=policy)
+        allocation_policy = batch_v1.AllocationPolicy(instances = [instances])
```

Comment:
May as well set the provisioning_models since you've added 'use_spot' to the schema? Or a TODO at least.

Reply:
[I think](https://cloud.google.com/python/docs/reference/batch/latest/google.cloud.batch_v1alpha.types.AllocationPolicy.InstancePolicy) you add something like `provisioning_model = ProvisioningModel.SPOT` to the `policy` object at line 344.

But fine to leave as TODO for now!

Reply:
Ah, got it. 🤦‍♀️  It's an attribute of InstancePolicy, whereas in the alpha version it's also an attribute of AllocationPolicy, which is where I was looking for it. Will add!